My log collection.

### **[mstsc.exe close [X] bottun ]**
[20210505-13:08:48] [DEBUG] [pid:1476 tid:140225398557120] Recieved disconnection request
[20210505-13:08:48] [ERROR] [pid:1476 tid:140225398557120] xrdp_sec_recv: xrdp_mcs_recv failed
[20210505-13:08:48] [ERROR] [pid:1476 tid:140225398557120] xrdp_rdp_recv: xrdp_sec_recv failed
[20210505-13:08:48] [ERROR] [pid:1476 tid:140225398557120] libxrdp_process_data: xrdp_rdp_recv failed
[20210505-13:08:48] [ERROR] [pid:1476 tid:140225398557120] xrdp_process_data_in: xrdp_process_loop failed
[20210505-13:08:48] [ERROR] [pid:1476 tid:140225398557120] SSL_write: I/O error
[20210505-13:08:48] [ERROR] [pid:1476 tid:140225398557120] xrdp_iso_send: trans_write_copy_s failed
[20210505-13:08:48] [ERROR] [pid:1476 tid:140225398557120] SSL_shutdown: Failure in SSL library (protocol error?)
[20210505-13:08:48] [DEBUG] [pid:1476 tid:140225398557120] Closed socket 11 (AF_INET 172.28.0.2:3389)
[20210505-13:08:48] [ERROR] [pid:1476 tid:140225398557120] Sending [ITU T.125] DisconnectProviderUltimatum failed
[20210505-13:08:48] [DEBUG] [pid:1476 tid:140225398557120] xrdp_mm_module_cleanup


### **[Target Win10 -> Start -> disconnect]**
ERRINFO_RPC_INITIATED_DISCONNECT_BY_USER (0x0000000B):
The disconnection was initiated by an administrative tool on the server running in the user's session.
rdp_recv_tpkt_pdu: disconnect
transport_check_fds: transport->recv_callback failed
transport_check_fds: do_callback failed
[20210505-13:02:04] [ERROR] [pid:1472 tid:140225398557120] lxrdp_check_wait_objs: freerdp_check_fds failed
[20210505-13:02:04] [DEBUG] [pid:1472 tid:140225398557120] Closed socket 11 (AF_INET 172.28.0.2:3389)
[20210505-13:02:04] [DEBUG] [pid:1472 tid:140225398557120] xrdp_mm_module_cleanup

### **[Target Win10 -> Start -> signout]**
ERRINFO_UNKNOWN 0x0000000C: Unknown error.
rdp_recv_tpkt_pdu: disconnect
transport_check_fds: transport->recv_callback failed
transport_check_fds: do_callback failed
[20210505-13:05:31] [ERROR] [pid:1474 tid:140225398557120] lxrdp_check_wait_objs: freerdp_check_fds failed
[20210505-13:05:31] [DEBUG] [pid:1474 tid:140225398557120] Closed socket 11 (AF_INET 172.28.0.2:3389)
[20210505-13:05:31] [DEBUG] [pid:1474 tid:140225398557120] xrdp_mm_module_cleanup

### **[Target Win10 -> Start -> reboot]**
ERRINFO_UNKNOWN 0x0000000C: Unknown error.
rdp_recv_tpkt_pdu: disconnect
transport_check_fds: transport->recv_callback failed
transport_check_fds: do_callback failed
[20210505-13:10:00] [ERROR] [pid:1477 tid:140225398557120] lxrdp_check_wait_objs: freerdp_check_fds failed
[20210505-13:10:00] [DEBUG] [pid:1477 tid:140225398557120] Closed socket 11 (AF_INET 172.28.0.2:3389)
[20210505-13:10:00] [DEBUG] [pid:1477 tid:140225398557120] xrdp_mm_module_cleanup

### **[another RDP connection estabulished]**
ERRINFO_DISCONNECTED_BY_OTHER_CONNECTION (0x00000005):
Another user connected to the server, forcing the disconnection of the current connection.
rdp_recv_tpkt_pdu: disconnect
transport_check_fds: transport->recv_callback failed
transport_check_fds: do_callback failed
[20210505-13:07:24] [ERROR] [pid:1475 tid:140225398557120] lxrdp_check_wait_objs: freerdp_check_fds failed
[20210505-13:07:24] [DEBUG] [pid:1475 tid:140225398557120] Closed socket 11 (AF_INET 172.28.0.2:3389)
[20210505-13:07:24] [DEBUG] [pid:1475 tid:140225398557120] xrdp_mm_module_cleanup

--- a/neutrinordp/xrdp-neutrinordp.c
+++ b/neutrinordp/xrdp-neutrinordp.c
@@ -182,18 +182,30 @@ lxrdp_connect(struct mod *mod)
         }
 
 #endif
-        LOG(LOG_LEVEL_ERROR, "freerdp_connect Failed to "
-            "destination :%s:%d",
+        LOG(LOG_LEVEL_ERROR, "NeutrinoRDP proxy connection: status [Failed]," 
+            " RDP client [%s:%s], RDP server [%s:%d], RDP server username [%s],"
+            " xrdp pamusername [%s], xrdp process id [%d]",
+            mod->client_info.client_addr,
+            mod->client_info.client_port,
             mod->inst->settings->hostname,
-            mod->inst->settings->port);
+            mod->inst->settings->port,
+            mod->inst->settings->username,
+            mod->pamusername,
+            g_getpid());
         return 1;
     }
     else
     {
-        LOG(LOG_LEVEL_INFO, "freerdp_connect returned Success to "
-            "destination :%s:%d",
+        LOG(LOG_LEVEL_INFO, "NeutrinoRDP proxy connection: status [Success],"
+            " RDP client [%s:%s], RDP server [%s:%d], RDP server username [%s],"
+            " xrdp pamusername [%s], xrdp process id [%d]",
+            mod->client_info.client_addr,
+            mod->client_info.client_port,
             mod->inst->settings->hostname,
-            mod->inst->settings->port);
+            mod->inst->settings->port,
+            mod->inst->settings->username,
+            mod->pamusername,
+            g_getpid());
     }
 
     return 0;
@@ -433,6 +445,16 @@ lxrdp_end(struct mod *mod)
     }
 
     LOG_DEVEL(LOG_LEVEL_DEBUG, "lxrdp_end:");
+    LOG(LOG_LEVEL_INFO, "NeutrinoRDP proxy connection: status [Disconnect],"
+        " RDP client [%s:%s], RDP server [%s:%d], RDP server username [%s],"
+        " xrdp pamusername [%s], xrdp process id [%d]",
+        mod->client_info.client_addr,
+        mod->client_info.client_port,
+        mod->inst->settings->hostname,
+        mod->inst->settings->port,
+        mod->inst->settings->username,
+        mod->pamusername,
+        g_getpid());
     return 0;
 }
 
@@ -504,9 +526,12 @@ lxrdp_set_param(struct mod *mod, const char *name, const char *value)
     {
         settings->desktop_resize = g_text2bool(value);
     }
-    else if (g_strcmp(name, "pamusername") == 0 ||
-             g_strcmp(name, "pampassword") == 0 ||
-             g_strcmp(name, "pammsessionmng") == 0)
+    else if (g_strcmp(name, "pamusername") == 0)
+    {
+        g_strncpy(mod->pamusername, value, 255);
+    }
+    else if (g_strcmp(name, "pampassword") == 0 ||
+             g_strcmp(name, "pamsessionmng") == 0)
     {
         /* Valid (but unused) parameters not logged */
     }

--- a/neutrinordp/xrdp-neutrinordp.h
+++ b/neutrinordp/xrdp-neutrinordp.h
@@ -216,5 +216,5 @@ struct mod
     struct bitmap_item bitmap_cache[4][4096];
     struct brush_item brush_cache[64];
     struct pointer_item pointer_cache[32];
-
+    char pamusername[255];
 };

--- a/xrdp/xrdp.ini.in
+++ b/xrdp/xrdp.ini.in
@@ -111,6 +111,12 @@ grey=dedede
 ls_top_window_bg_color=009cb5
 
 ; width and height of login screen
+;
+; The default height allows for about 5 fields to be comfortably displayed
+; above the buttons at the bottom. To display more fields, make <ls_height>
+; larger, and also increase <ls_btn_ok_y_pos> and <ls_btn_cancel_y_pos>
+; below
+;
 ls_width=350
 ls_height=430
 
@@ -216,6 +222,9 @@ port=-1
 ; display number of the session
 #chansrvport=DISPLAY(0)
 
+; Generic VNC Proxy
+; Tailor this to specific hosts and VNC instances by specifying an ip
+; and port and setting a suitable name.
 [vnc-any]
 name=vnc-any
 lib=libvnc.@lib_extension@
@@ -228,13 +237,23 @@ password=ask
 #pamsessionmng=127.0.0.1
 #delay_ms=2000
 
+; Generic RDP proxy using NeutrinoRDP
+; Tailor this to specific hosts by specifying an ip and port and setting
+; a suitable name.
 [neutrinordp-any]
 name=neutrinordp-any
+; To use this section, you should build xrdp with configure option
+; --enable-neutrinordp.
 lib=libxrdpneutrinordp.@lib_extension@
 ip=ask
 port=ask3389
 username=ask
 password=ask
+; Uncomment the following lines to enable PAM authentication for proxy
+; connections.
+#pamusername=ask
+#pampassword=ask
+#pamsessionmng=127.0.0.1
 ; Currently NeutrinoRDP doesn't support dynamic resizing. Uncomment
 ; this line if you're using a client which does.
 #enable_dynamic_resizing=false

